File: src/tools/editor.ts
import * as fs from "fs";
import * as path from "path";
import { Workspace } from "../workspace.js";

export class CodeEditor {
  constructor(private workspace: Workspace) {}

  readFile(filepath: string): string {
    return fs.readFileSync(this.workspace.resolve(filepath, true), "utf-8");
  }

  writeFile(filepath: string, content: string): void {
    const resolved = this.workspace.resolve(filepath, false);
    const dir = path.dirname(resolved);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    fs.writeFileSync(resolved, content, { encoding: "utf-8", flag: "wx" });
  }

  /**
   * Applies a patch using exact block replacement.
   * This is safer than a global string replacement.
   */
  applyPatch(filepath: string, searchBlock: string, replaceBlock: string): void {
    const resolved = this.workspace.resolve(filepath, true);
    const content = fs.readFileSync(resolved, "utf-8");
    const occurrences = content.split(searchBlock).length - 1;
    if (occurrences !== 1) {
      throw new Error(
        `Apply patch to ${JSON.stringify(resolved)} expected exactly one search block, found ${occurrences}.`
      );
    }
    const newContent = content.replace(searchBlock, replaceBlock);
    fs.writeFileSync(resolved, newContent, "utf-8");
  }

  listFiles(dir: string): string[] {
    const resolvedDirectory = this.workspace.resolve(dir, true);
    let results: string[] = [];
    const list = fs.readdirSync(resolvedDirectory);
    for (const file of list) {
      if (file === 'node_modules' || file === '.git' || file === 'dist') continue;
      const fullPath = path.join(resolvedDirectory, file);
      const stat = fs.lstatSync(fullPath);
      if (stat.isSymbolicLink()) continue;
      if (stat && stat.isDirectory()) {
        results = results.concat(this.listFiles(fullPath));
      } else {
        results.push(fullPath);
      }
    }
    return results;
  }
}

File: editor/editor.go
package editor

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/donomii/j/types"
	"github.com/donomii/j/workspace"
)

type CodeEditor struct {
	workspaceRoot types.WorkspaceRoot
}

func New(workspaceRoot types.WorkspaceRoot) *CodeEditor {
	return &CodeEditor{workspaceRoot: workspaceRoot}
}

func (e *CodeEditor) ReadFile(path string) (string, error) {
	resolved, err := workspace.Resolve(e.workspaceRoot, path, true)
	if err != nil {
		return "", fmt.Errorf("read file %q: %w", path, err)
	}
	data, err := os.ReadFile(resolved)
	return string(data), err
}

func (e *CodeEditor) WriteFile(path, content string) error {
	resolved, err := workspace.Resolve(e.workspaceRoot, path, false)
	if err != nil {
		return fmt.Errorf("create file %q: %w", path, err)
	}
	if err := os.MkdirAll(filepath.Dir(resolved), 0755); err != nil {
		return fmt.Errorf("create parent directory for %q: %w", resolved, err)
	}
	file, err := os.OpenFile(resolved, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
	if err != nil {
		return fmt.Errorf("create new file %q without replacing existing data: %w", resolved, err)
	}
	if _, err := file.WriteString(content); err != nil {
		file.Close()
		return fmt.Errorf("write new file %q: %w", resolved, err)
	}
	if err := file.Close(); err != nil {
		return fmt.Errorf("close new file %q: %w", resolved, err)
	}
	return nil
}

func (e *CodeEditor) ApplyPatch(path, search, replace string) error {
	resolved, err := workspace.Resolve(e.workspaceRoot, path, true)
	if err != nil {
		return fmt.Errorf("apply patch to %q: %w", path, err)
	}
	data, err := os.ReadFile(resolved)
	if err != nil {
		return fmt.Errorf("read patch target %q: %w", resolved, err)
	}
	content := string(data)
	count := strings.Count(content, search)
	if count != 1 {
		return fmt.Errorf("apply patch to %q: expected exactly one search block, found %d", resolved, count)
	}
	if err := os.WriteFile(resolved, []byte(strings.Replace(content, search, replace, 1)), 0644); err != nil {
		return fmt.Errorf("write patched file %q: %w", resolved, err)
	}
	return nil
}

func (e *CodeEditor) ListFiles(dir string) ([]string, error) {
	resolved, err := workspace.Resolve(e.workspaceRoot, dir, true)
	if err != nil {
		return nil, fmt.Errorf("list files in %q: %w", dir, err)
	}
	skip := map[string]bool{"node_modules": true, ".git": true, "dist": true}
	var results []string
	err = filepath.Walk(resolved, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.Mode()&os.ModeSymlink != 0 {
			return nil
		}
		if info.IsDir() && skip[info.Name()] {
			return filepath.SkipDir
		}
		if !info.IsDir() {
			results = append(results, path)
		}
		return nil
	})
	return results, err
}
